monitors/CRMonitor.js

```diff
@@ -0,0 +1,135 @@
+  //  [{ name: "Market Making bot",
+  //    address: "0x12345",
+  //    crAlert: 150 },
+  // ...];
+  constructor(logger, expiringMultiPartyClient, walletsToMonitor) {
+    this.logger = logger;
+
+    this.empClient = expiringMultiPartyClient;
+    this.empContract = this.empClient.emp;
+    this.web3 = this.empClient.web3;
+
+    this.walletsToMonitor = walletsToMonitor;
+
+    this.formatDecimalString = createFormatFunction(this.web3, 2);
+
+    // TODO: replace this with a fetcher that pulls the actual collateral token symbol
+    this.collateralCurrencySymbol = "DAI";
+    this.syntheticCurrencySymbol = "UMATEST";
+  }
+
+  // Queries all monitored wallet ballance for collateralization ratio against a given threshold
+  checkWalletCrRatio = async priceFunction => {
+    // yield the price feed at the current time.
+    const contractTime = await this.empContract.methods.getCurrentTime().call();
+    const priceFeed = priceFunction(contractTime);
+    this.logger.debug({
+      at: "CRMonitor",
+      message: "Checking wallet collateralization ratios",
+      price: priceFeed
+    });
+    // For each monitored wallet check if the current collaterlization ratio is below the monitored threshold.
+    // If it is, then send an alert of formatted markdown text.
+    for (let wallet of this.walletsToMonitor) {
+      const { shouldPush, crRatio } = this._shouldPushWalletNotification(wallet, priceFeed);
+      if (shouldPush) {
+        // Sample message:
+        // Risk alert: [Tracked wallet name] has fallen below [threshold]%.
+        // Current [name of identifier] value: [current identifier value].
+        const mrkdwn =
+          wallet.name +
+          " (" +
+          createEtherscanLinkMarkdown(this.web3, wallet.address) +
+          ") collateralization ratio has dropped to " +
+          this.formatDecimalString(crRatio) +
+          "% which is below the " +
+          wallet.crAlert +
+          "% threshold. Current value of " +
+          this.syntheticCurrencySymbol +
+          " is " +
+          this.formatDecimalString(priceFeed);
+
+        this.logger.info({
+          at: "ContractMonitor",
+          message: "Collateralization ratio alert 🚨!",
+          mrkdwn: mrkdwn
+        });
+      }
+    }
+  };
+
+  _getPositionInformation = address => {
+    const positionInfo = this.empClient.getAllPositions().filter(position => position.sponsor == address);
+    if (positionInfo.length == 0) {
+      return null;
+      // there should only ever be one position information object per address
+    } else return positionInfo[0];
+  };
+
+  _shouldPushWalletNotification(wallet, priceFeed) {
+    const positionInformation = this._getPositionInformation(wallet.address);
+    if (positionInformation == null) {
+      // There is no position information for the given wallet.
+      return [false, 0];
```

Comment:
Instead of having to return a tuple `(shouldPush, CR)`, have you considered  returning `-1` or `null` for errors?

Reply:
sure. this was more complex before with the rate-limiting. I can return the CR value and if this is larger than 0 then it will send the message. This'll make it far cleaner.